smarts/core/traffic_history.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (C) 2020. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+
+from cached_property import cached_property
+from contextlib import nullcontext, closing
+from functools import lru_cache
+import random
+import sqlite3
+
+
+class TrafficHistory:
+    def __init__(self, db):
+        self._db = db
+        self._db_cnxn = None
+
+    def connect_if_not(self):
+        if not self._db_cnxn:
+            self._db_cnxn = sqlite3.connect(self._db)
+
+    def close_db(self):
+        if self._db_cnxn:
+            self._db_cnxn.close()
+            self._db_cnxn = None
+
+    def _query_val(self, query, params=(), result_type=float):
```

Comment:
I think a default `result_type` is dangerous here. A user should always know what type they are going to get back.

Reply:
Yeah, agreed -- the user should always know, although the problem was that then `params` cannot be optional either, and I think it's important for it to be, and I also think that re-ordering `params` to be the last argument is not desirable since it's sort of "bound more tightly" to the `query` argument.

I guess it could be the very first argument (sort of like a templatized function in C++).  I'll try that and see how it looks...

Reply:
OK, I think I'm ok with the way it looks as the first argument.  Going with that.